plugins/trip.hpp

```diff
@@ -0,0 +1,332 @@
+        BOOST_ASSERT_MSG(result_table.size() > 0, "Distance Table is empty.");
+
+        // get scc components
+        SCC_Component scc = [&](){
+            if (*std::max_element(result_table.begin(), result_table.end()) == INVALID_EDGE_WEIGHT) {
+                // compute all scc with tarjan
+                return SplitUnaccessibleLocations(number_of_locations, result_table);
+            } else {
+                // whole graph is one scc
+                std::vector<NodeID> location_ids(number_of_locations);
+                std::iota(std::begin(location_ids), std::end(location_ids), 0);
+                return SCC_Component(location_ids);
+            }
+        }();
+
+
+        using NodeIDIterator = typename std::vector<NodeID>::const_iterator;
+
+        std::vector<std::vector<NodeID>> route_result;
+        route_result.reserve(scc.GetNumberOfComponents());
+        TIMER_START(trip);
+        //run Trip computation for every SCC
+        for (std::size_t k = 0; k < scc.GetNumberOfComponents(); ++k) {
+            const auto component_size = scc.range[k+1] - scc.range[k];
+
+            BOOST_ASSERT_MSG(component_size >= 0,"invalid component size");
+
+            if (component_size > 1) {
+                std::vector<NodeID> scc_route;
+                NodeIDIterator start = std::begin(scc.component) + scc.range[k];
+                NodeIDIterator end = std::begin(scc.component) + scc.range[k+1];
+
+                if (component_size < BF_MAX_FEASABLE) {
+                    scc_route = osrm::trip::BruteForceTrip(start, end, number_of_locations, result_table);
+                    route_result.push_back(scc_route);
+                } else {
+                    scc_route = osrm::trip::FarthestInsertionTrip(start, end, number_of_locations, result_table);
+                    route_result.push_back(scc_route);
+                }
+
+                // use this if output if debugging of route is needed:
+                SimpleLogger().Write() << "Route #"
+                                       << k << ": "
+                                       << [&scc_route](){
+                                               std::string s = "";
+                                               for (auto x : scc_route) {
+                                                   s += std::to_string(x) + " ";
+                                               }
+                                               return s;
+                                           }();
+            } else {
+                // if component only consists of one node, add it to the result routes
+                route_result.push_back({scc.component[scc.range[k]]});
+            }
+        }
+
+        // compute all round trip routes
+        std::vector<InternalRouteResult> comp_route (route_result.size());
+        for (std::size_t r = 0; r < route_result.size(); ++r) {
+            ComputeRoute(phantom_node_vector, route_parameters, route_result[r], comp_route[r]);
```

Comment:
Instead of out-references better return the min. route by value (that is, move it out) in `ComputeRoute`.

This way, you can use something like this here:

    std::transform(begin(route_result), end(route_result), begin(comp_route), getRoute);

with `getRoute` being a lambda, like:

    auto getRoute = [&] (const auto& route_result) {
      return ComputeRoute(phantom_node_vector, route_parameters, route_result);
    };